go-controller/pkg/ovn/logical_switch_manager/logical_switch_manager.go

```diff
@@ -321,6 +321,39 @@ func (manager *LogicalSwitchManager) ReleaseIPs(nodeName string, ipnets []*net.I
 	return nil
 }
 
+// ConditionalIPRelease determines if any IP is available to be released from an IPAM conditionally if func is true.
+// It guarantees state of the allocator will not change while executing the predicate function
+func (manager *LogicalSwitchManager) ConditionalIPRelease(nodeName string, ipnets []*net.IPNet, predicate func() (bool, error)) (bool, error) {
+	manager.RLock()
+	defer manager.RUnlock()
+	if ipnets == nil || nodeName == "" {
+		klog.V(5).Infof("Node name is empty or ip slice to release is nil")
+		return false, nil
+	}
+	lsi, ok := manager.cache[nodeName]
+	if !ok {
+		return false, nil
+	}
+	if len(lsi.ipams) == 0 {
+		return false, nil
+	}
+
+	// check if ipam has one of the ip addresses, and then execute the predicate function to determine
+	// if this IP should be released or not
+	for _, ipnet := range ipnets {
+		for _, ipam := range lsi.ipams {
+			cidr := ipam.CIDR()
+			if cidr.Contains(ipnet.IP) {
+				if ipam.Has(ipnet.IP) {
+					return predicate()
```

Comment:
I might be overthinking this... but here we are basically returning the result of `predicate()` as soon as we see that an IP address of this pod is in the node IPAM. `predicate` doesn't take any input arguments and returns false if any of the pod IPs (0 or 1 IPv4 + 0 or 1 IPv6) is found in any other non-completed pods of the same node. 

When it's single stack it all works fine, there's only one IP address to consider, but since the pod might be also dual stack, should we actually pass `ipnet` (that is, the pod IP we're considering) to `predicate`?

For instance, let's say that a our pod A is dual stack, it has IP1 and IP2.
- IP1 (v4): is in node IPAM, is still assigned only to pod A
- IP2 (v6): is in node IPAM, but has been reassigned to another pod.

`ConditionalIPRelease` will process `IP1` first, it will see it is in the node IPAM, it will call `predicate()`; `predicate()` will check *both* IPs of pod A against the IPs of all other pods on the node and will see that IP2 belongs to another pod, so `predicate()` will return false. We were checking if we can release `IP1`, but we got a result for `IP2`. The result is that `IP1` will never be released, unless the other pod that has `IP2` gets deleted at some point.

Is this something to worry about or can we safely ignore it?

Reply:
this smells like legit problem... we should fix? good catch

Reply:
oh wait this won't be a problem right? Because Tim's predicate function is actually doing:
```
for _, pAddr := range pAddrs {
	for _, podAddr := range podIfAddrs {
	    if pAddr.Equal(podAddr.IP) {
		klog.Infof("Will not release IP address: %s for pod %s/%s. Detected another pod"+
			" using this IP: %s/%s", pAddr.String(), pod.Namespace, pod.Name, p.Namespace, p.Name)
	        return false, nil
	    }
         }
}
klog.Infof("Releasing IPs for Completed pod: %s/%s, ips: %s", pod.Namespace, pod.Name,
				util.JoinIPNetIPs(podIfAddrs, " "))
return true, nil
```

He sets shouldRelease = true only if all IPs of this pod are not being used by any other pod? Of course the reverse could be a problem where one IPV4 is ready for release but we don't release it because the IPV6 one is held by another pod, but maybe that's ok.